### ISSUE TITLE
Derma Abuse

### ISSUE THREAD

**Issue by edgarasf123** [Facepunch/garrysmod-issues#1911]
https://dl.dropboxusercontent.com/u/8081284/ShareX/2015/03/2015-03-20_09-21-28.mp4
Some servers are using derma to hide ESC menu, in this case I was afk on some server and it forced to "AFK Server" which doesn't allow to use ESC. The only way to leave the server(without joining to their other servers) was to use console. But not everyone knows how to use console.

**Comment by Noiwex** [comment]
It's a feature.

**Comment by edgarasf123** [comment]
How is this a feature? Servers abuse derma and it's a problem which should be resolved, not a perk.

**Comment by Noiwex** [comment]
Then don't visit that server.

**Comment by thegrb93** [comment]
>Then don't visit that server.

That's not a solution.

**Comment by Noiwex** [comment]
There's no ways to solve this other than breaking everything. It's not a
bug, again.

**Comment by thegrb93** [comment]
Here's a nice example.
```

local HTML = [[
<html>
<head>
	<title>BSoD</title>
</head>

<body bgcolor="#000088" scroll="no">
	<font face="Lucida Console" size="3" color="#FFFFFF">
	<p>A problem has been detected and Lua has been shutdown to prevent damage to your carpet.</p>
	
	<p>The problem seems to be caused by the following player: %s</p>
	
	<p>ERR_PLAYER_IS_DIRTBAG</p>
	<br>
	
	<p>If this is the first time you've seen this stop error screen, restart your GMod. If this screen appears again, follow these steps:</p>
	
	<p>Check to make sure you're not running any cheats or hacks. This includes even having the files. If this is a new installation, disable or remove any newly installed addons.</p> 
	
	<p>If problems continue. Clean your GMod by deleting the "garrysmod" folder at "%s", then rejoin. This will download a fresh one.</p> 
	
	<p>Technical information:</p> 
	
	<p>*** STOP: 0x0000541D (0x0000000C,0x00000002,0x00000000,0xF86B5A89)</p> 
	<p>***       %s - Address %s base at ILuaBase, DateStamp %s</p>
	<br>
	
	<p>Beginning dump of physical trousers..</p> 
	<p>Dump complete.</p> 
	<p>Contact HeX (http://steamcommunity.com/id/MFSiNC) for further assistance.</p> 
	
	</font> 
</body>
</html>
]]


local BSoD = nil

local function MakeBluescreen(um)
	local Show = um:ReadBool()
	
	if BSoD and not Show then
		
		BSoD:Remove()
		BSoD = nil
		return
	end
	
	HTML = Format(HTML,
		LocalPlayer():Nick(),
		util.RelativePathToFull("gameinfo.txt"):gsub("gameinfo.txt",""):Trim("\\"),
		LocalPlayer():SteamID(),
		tostring( LocalPlayer().SteamID ),
		os.time()
	)
	
	BSoD = vgui.Create("HTML")
		BSoD:SetPos(0,0)
		BSoD:SetSize( ScrW(), ScrH() )
	BSoD:SetHTML(HTML)
end
usermessage.Hook("BSoD", MakeBluescreen)


local function BSoDBinds(ply,cmd,down)
	if BSoD then return true end
end
hook.Add("PlayerBindPress", "BSoDBinds", BSoDBinds)

local function BSoDThink()
	if BSoD then gui.HideGameUI() end
end
hook.Add("PreDrawHUD", "BSoDThink", BSoDThink)
```
https://github.com/MFSiNC/HAC/blob/f1f8159f1ce27c175215a545fbecebec34a2d0b9/v13/HeX%27s%20AntiCheat/lua/HAC/cl_BSoD.lua

**Comment by wiox** [comment]
> We know it's not a bug, but it's a feature that needs rethinking.

I agree. Sadly, it's unlikely to change. I suggest you bind a key to disconnect (but servers can even stop that).

**Comment by edgarasf123** [comment]
How is fixing this going to break anything. Just prioritize the ESC window on top of any other vgui windows.

**Comment by wiox** [comment]
Some servers use this legitimately.

**Comment by edgarasf123** [comment]
Well some servers used running cmd commands legitimately(like opening websites with OS browser), other used it to run malicious files. Does that mean just because someone use it legitimately we should ignore malicious uses?

**Comment by wiox** [comment]
I'm saying it's unlikely to be fixed. If anybody cared about this the game.HideGameUI wouldn't have been added in the first place.

**Comment by X-Coder** [comment]
There is a lot in lua which could be used to do malicious stuff, we just should disable lua all together, just to be safe.

If a server owner does it on purpose without any valid reason, don't visit his server again.

**Comment by neico** [comment]
"Blacklist server" anyone? Valve added this for a reason...

**Comment by XenMix** [comment]
Press your "console" key, then you go back to Gmod menu...

**Comment by thegrb93** [comment]
>There is a lot in lua which could be used to do malicious stuff

Anything that can effectively "crash" someone should be fixed. This is one such case.

**Comment by X-Coder** [comment]
thegrb93, you are talking about killing the game, because the main menu is hidden = not a crash problem.

Of course should anything that causes crashes be fixed, but I don't see where edgarasf123 or you have a crash problem, he was talking about a hidden main menu.

Outside the game, I guess some server owners need the HideGameUI function for example in the loading screen to hide the default stuff. But whats the reason to allow hiding the main menu and console after the loading screen finished?

**Comment by thegrb93** [comment]
I said it effectively crashes the game because you have no alternative at that point other than to kill the process. I think the correct term is "soft lock".

**Comment by neico** [comment]
Then you aren't aware of the correct definition of a "crash".
A crash is when the program automatically closes itself because of a fatal error it can't recover from. (usually your OS gives you a pop-up to either let it close or search for the issue, or send the crash dump to the developers)

Your issue get's near the definition of a "freeze" where the program is still running, but is not responding and displaying anything anymore.

But in the end it's just a feature that can be abused (like mostly anything, gmod is a sandbox after all, it's not affecting your entire computer, not even the entire game, just some idiotic servers that abuse it).

What could be done to deescalate this a bit would be a shortcut to recover the GameUI, something like CTRL + M or so... (maybe with a little notification panel on the left corner of the screen once the GameUI is hidden)

**Comment by thegrb93** [comment]
I already fucking explained myself dude. Look up what it means when someone says something "effectively" happens. It's the effect of a crash even though it isn't a crash.

**Comment by Noiwex** [comment]
Wow it's fucking nothing

**Comment by Kefta** [comment]
I know multiple servers that use this legitimately; I'm pretty sure there's even some custom escape menus on script fodder. Really, there should be a clientside convar that servers couldn't override to hide these escape menus

**Comment by TheFreezebug** [comment]
Its not broken so it need not be fixed.  Either way, if you really wanted to disconnect you could always use the chat escape for console commands by typing the following ' ";disconnect '

**Comment by wiox** [comment]
You can't any more.
